src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs

```diff
@@ -1361,6 +1362,64 @@ public static bool IsFileTypeOrUsesFileTypes(this TypeSymbol type)
             return foundType is not null;
         }
 
+        internal static string? AssociatedFileIdentifier(this NamedTypeSymbol type)
+        {
+            if (type is not SourceMemberContainerTypeSymbol { AssociatedSyntaxTree: SyntaxTree tree })
+            {
+                return null;
+            }
+            var ordinal = type.DeclaringCompilation.GetSyntaxTreeOrdinal(tree);
```

Comment:
>var ordinal = type.DeclaringCompilation.GetSyntaxTreeOrdinal(tree); [](http://example.com/codeflow?start=12&length=67)

What's the purpose of the ordinal? Including the ordinal will make EnC harder to implement.

Reply:
Is it to make the name unique? I think we need another approach for that that does not depend on syntax tree ordering.

Reply:
One way would be to use normalized full path (after applying source map) and calculate SHA2 hash from it.
I don't see much value in including the file name in the generated type name.

Reply:
Yeah. This was done under the assumption that it's possible to create a compilation using syntax trees with the exact same path, and even the exact same content.

The file element in the PDB contains an `id` field which seems to correspond to this ordinal. Is that not the case?

Reply:
I think the compiler now orders the documents in the PDB such that their ordinal corresponds to syntax tree ordinals and places documents that do not have corresponding syntax tree (created for `#line` directives) last. So that should not be a problem. 

However, let's say you have compilation with files [a.cs, b.cs, c.cs] and you delete b.cs from the compilation during EnC. The new compilation will then have [a.cs, c.cs] and the ordinal for file type definition in c.cs will change and thus the name of the file. That's a problem because now this change caused existing type in c.cs to be renamed, which is not supported by EnC. To avoid that the IDE would need to keep track of the files over the debugging session and then pass the compiler a map to use to map the new ordinals to the previous ordinals. It is possible to do so but adds extra layers of complexity in multiple places.

I'd very much prefer not to do that and instead use SHA256 instead of the ordinal.

> This was done under the assumption that it's possible to create a compilation using syntax trees with the exact same path

Is it possible? I'm almost sure it's not possible from command line. Do we not check for duplicate paths when emitting compilation? That seems like a scenario I'd gladly not support/break. It's not clear to me how the PDB document table even works in that case.

Reply:
I did find that compilations from the command line don't support passing multiple files from the same path, FWIW. The extra instances of the "same" file are discarded. However, compilations made via the APIs frequently have empty strings or duplicate strings as paths--at least in our tests.

This documentation on SyntaxTree.FilePath also concerned me:

https://github.com/dotnet/roslyn/blob/3d735b9a32d37934269eefe209dcc0fec3e46020/src/Compilers/Core/Portable/Syntax/SyntaxTree.cs#L44-L45

Reply:
We should report emit error instead of ignoring the file. The compilation can still allow duplicate file names, but it shouldn't allow emit.